clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -327,6 +345,17 @@ public void poll(long timeout) throws IOException {
 
         addToCompletedReceives();
 
+        // finish connection setup
+        if (connectableChannels.size() > 0) {
+            for (KafkaChannel channel : connectableChannels) {
+                channel.finishConnect();
```

Comment:
This can be tricky since finishConnect() can throw an IOException. This is dealt with in the while loop starting at line 293. We can deal with it here again, but we will end up duplicating more and more code. Have you considered iterating both keys from selector and immediately connected keys together? That complicates the iteration a bit, but will avoid code duplication.